### DISCUSSION TITLE
Example for react toggle

### DISCUSSION THREAD

**Discussion by razzeee** [floating-ui/floating-ui#1467] [General]
Are there any examples for a toggle tooltip with react?

**Comment by atomiks** [comment]
https://floating-ui.com/docs/react-dom#usage

In this example would just add a piece of state, `const [open, setOpen] = useState(false)` and use that to conditionally render the tooltip. Add `onMouseEnter`, `onMouseLeave` etc. event listeners on the reference element to change that state.

**Reply by razzeee** [reply]
Thank you, get this going after your post. But ended up not getting the whole thing to work. Ended up just using css with a fixed position.

**Reply by atomiks** [reply]
What did you have trouble with? FYI this package only handles positioning logic right now, but we're working on something that can craft floating elements like a tooltip more easily too, so it won't be as time consuming to implement them from scratch.

**Reply by razzeee** [reply]
Basically the tooltip shown in this screenshot https://github.com/flathub/frontend/pull/128 
With this library it ended up just centered on the `info` icon. Tried autoPlacement and shift maybe even more, not sure if that's a problem with SSG/nest.js

**Comment by silouanwright** [comment]
A couple of interesting things I noticed to be aware of (for anyone else stumbling across this):

- The React examples on https://floating-ui.com/ are not using floating-ui at all. They're using Tippy, the older library.
- The only thing that has been released for React is the positioning utils.
- Further than that, there is no provided examples or documentation to spin this up into a workable Tooltip. You've tasked with figuring that out (React state, necessary event listeners), however most people are "probably" not going to want to do that, unless you want to create a bunch of functionality the components will likely support anyway, that the old headless Tippy supported (delays, animations, etc etc).
- @atomiks created a super secret component here: https://github.com/floating-ui/floating-ui/blob/master/website/lib/components/Floating.js which I imagine will start forming the underlying basis of the components that end up being created, that "is" using floating-ui react dom positioning utils
- I replaced a Popper version of my Tooltip with floating-ui: https://github.com/cbinsights/form-design-system/pull/972/files, and here's a codesandbox that generated from this PR, and you can see that Tooltip works: https://codesandbox.io/s/cbinsights-form-design-system-forked-wim8x
- Again, I don't personally want to go and implement a lot of the features I assume @atomiks is going to provide in the component facade, so I'm going to wait a little until the components get created and fleshed out.

**Reply by atomiks** [reply]
- Haha yeah I was working on the docs before the new package was finished, so Tippy was used everywhere initially on the docs (actually, the docs were originally going to be Tippy v7 but I pivoted). It's left there now due to being lazy and using something that has animations already without adding a big spring library or dealing with CSS/unmounting.
- So far, it's the next evolution of Popper, exporting Tippy's work under `@floating-ui` is the next step. I'm still exploring how exactly this should be done. Just something like Radix primitives, `@floating-ui/tooltip`, or a generic `@floating-ui/generic` that's more unopinionated like Tippy, but less accessible out of the box? Or maybe composable hooks/pieces like `useTrigger()`? I've been exploring it for a while and it's hard to decide exactly what the best course of action is.
- I figured the tutorial would be enough for React devs to be able to translate into something. All it requires is a boolean piece of state and some event listeners which (to me at least) seems like it would come naturally, but who knows
- That god component is VeryBad™ and is basically one giant antipattern hack that I spun up to be able to quickly make all the demos inline in MDX. I wouldn't recommend anything like that in real code 😄 
- —
- If you need the features now, feel free to make them. But yeah all of the use cases that popped up from Tippy over the years will be supported, just in a better way based on what I've learned.

**Reply by silouanwright** [reply]
Good to know!

>  I've been exploring it for a while and it's hard to decide exactly what the best course of action is.

I feel like what might be really nice, is to start working backwards from the things people want from a functionality / customizability perspective. It might be clearer, after mapping the common use cases, the usual path of feature requests, what the best API for that is? And also looping in the community for their opinion, even if you don't ultimately go with the popular choice. We've been subjected to a lot of interesting library patterns.

> I figured the tutorial would be enough for React devs to be able to translate into something. All it requires is a boolean piece of state and some event listeners which (to me at least) seems like it would come naturally, but who knows

Yeah, it is for sure. I'm just a bit spoiled knowing how good tippy/headless is, and developing feature parity with whatever you're going to complete is a lost cause, so I'll wait for you 🤣

**Reply by atomiks** [reply]
Here's my current vision for the components package: https://twitter.com/atomiksdev/status/1483686942223253507

The 3 core components that I've used this strategy on are a tooltip, popover, and menu dropdown — the composable effects can build any of them.

The beauty of this is that you can build any UI component using these primitive building blocks, even for ones which there's not yet built-in support like the logic to write a `<Select />`. You don't need to wait for this library to "officially" support it — create your own effects to do the work (like a custom positioning middleware). 

Also it's not locked into React, so it will work with any UI lib or framework. I think I'll keep it locked to the DOM as a platform at least initially though, so it'll probably be called `@floating-ui/dom-components` or such
